osu.Game/Screens/Play/SongProgress.cs

```diff
@@ -140,6 +139,52 @@ protected override void LoadComplete()
 
             AllowSeeking.BindValueChanged(_ => updateBarVisibility(), true);
             ShowGraph.BindValueChanged(_ => updateGraphVisibility(), true);
+
+            migrateSettingFromConfig();
+        }
+
+        [Resolved]
+        private OsuConfigManager config { get; set; }
+
+        [Resolved]
+        private SkinManager skinManager { get; set; }
+
+        /// <summary>
+        /// This setting has been migrated to a per-component level.
+        /// Only take the value from the config if it is in a non-default state (then reset it to default so it only applies once).
+        ///
+        /// Can be removed 20221027.
+        /// </summary>
+        private void migrateSettingFromConfig()
+        {
+            Bindable<bool> configShowGraph = config.GetBindable<bool>(OsuSetting.ShowProgressGraph);
+
+            if (!configShowGraph.IsDefault)
+            {
+                ShowGraph.Value = configShowGraph.Value;
+                configShowGraph.SetDefault();
+
+                // This is pretty ugly, but the only way to make this stick...
+                if (skinManager != null)
+                {
+                    var skinnableTarget = this.FindClosestParent<ISkinnableTarget>();
+
+                    if (skinnableTarget != null)
+                    {
+                        skinManager.EnsureMutableSkin();
+
+                        // If `EnsureMutableSkin` actually changed the skin, default layout may take a frame to apply.
+                        // See `SkinnableTargetComponentsContainer`'s use of ScheduleAfterChildren.
+                        ScheduleAfterChildren(() =>
+                        {
+                            var skin = skinManager.CurrentSkin.Value;
+                            skin.UpdateDrawableTarget(skinnableTarget);
+
+                            skinManager.Save(skin);
+                        });
+                    }
+                }
```

Comment:
This won't exactly work if the skin is not mutable, since reloading the skin would actually clear all the current components (including this one) and reloading to the ones from the new skin (mutable skin).

`SongProgress` should instead bail out from migrating if the current skin is not mutable, and leaving it for the new instance to migrate:
```suggestion
                // if the current skin is not mutable, reload to a mutable variant
                // and early-return for the migration to occur on the mutable skin.
                if (skinManager.CurrentSkinInfo.Value.PerformRead(p => p.Protected))
                {
                    skinManager.EnsureMutableSkin();
                    return;
                }

                ShowGraph.Value = configShowGraph.Value;

                var skinnableTarget = this.FindClosestParent<ISkinnableTarget>();

                // This is pretty ugly, but the only way to make this stick...
                if (skinnableTarget != null)
                {
                    var skin = skinManager.CurrentSkin.Value;
                    skin.UpdateDrawableTarget(skinnableTarget);
                    skinManager.Save(skin);
                }

                configShowGraph.SetDefault();
```

I'm definitely unsure on the usage of `PerformRead` to check whether the skin is mutable or not, however. I guess it's fine because we're in the update thread..?

Reply:
Your solution looks like a good path, although probably better to add a return `bool` from `EnsureMutableSkin` instead of doing the check manually.

Reply:
I've updated the code. Looks like we can omit the schedule as well, so that's a plus.

Reply:
Yeah turns out scheduling is still required, because calling `UpdateDrawableTarget` creates skinnable info of the current target state, overriding the [default HUD layout](https://github.com/ppy/osu/blob/26ecf60f1fcd0480a1536144508b64ceecf843ed/osu.Game/Skinning/DefaultSkin.cs#L82-L155). Therefore `UpdateDrawableTarget` needs to be called after layout took place for the components to have their positions serialised.

I feel this can be improved in different ways, but probably fine for now given it's temporary migration logic.